release-23.2: catalog/lease: optimize lease count query

fqazi opened the pull request:
backport of 4f7e3944acae753b8b3a23283aadf247556eb7d9

Previously, the query to count leases would end up scanning the entire system.leases table because of the addition of crdb_region as the first column on 23.1. This meant that scans not specifying a region would always be forced to scan the entire table. To address this, this patch will update the count query to specify the default region when working on single region system databases.

Fixes: #119456

Release note (bug fix): Schema changes with a large number of descriptors could end up doing full table scans on system.leases.



Release justification: a low risk fix that optimizes the leasing query for non-multi-region clusters, which was causing a regression for customers.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119464)
<!-- Reviewable:end -->

fqazi committed:
release-23.2: catalog/lease: optimize lease count query

Previously, the query to count leases would end up scanning the entire
system.leases table because of the addition of crdb_region as the first
column on 23.1. This meant that scans not specifying a region would
always be forced to scan the entire table. To address this, this patch
will update the count query to specify the default region when working
on single region system databases.

Fixes: #119456

Release note (bug fix): Schema changes with a large number of descriptors
could end up doing full table scans on system.leases.

rimadeodhar reviewed (changes requested):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119464)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/catalog/descs/txn.go` line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrIRapG8l6KDv5GzBoc:-NrIRapG8l6KDv5GzBod:bm8fdo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/descs/txn.go#L71)):*
> ```Go
> 
> 	// Get a lease on the system database descriptor to determine if we
> 	// should to multi-region leasing queries.
> ```

Nit: incomplete comment.
___
*[`pkg/sql/catalog/lease/count.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrIRktP8Ro4K1PxzoFQ:-NrIRktP8Ro4K1PxzoFR:b-1ooiuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/count.go#L46)):*
> ```Go
> 	// to use the default region.
> 	querySingleRegion := " AND crdb_region=$2"
> 	if !isMultiRegion {
> ```

Shouldn't this be `isMultiRegion` instead of `!isMultiRegion`? We need to add the `crdb_region` clause only for single region clusters right?
___
*[`pkg/sql/catalog/lease/lease.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrIRwq717upG601cPDi:-NrIRwq717upG601cPDj:b99n5w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L88)):*
> ```Go
> 	}
> 	querySingleRegion := ""
> 	if desc, ok := sysDBDesc.Underlying().(catalog.DatabaseDescriptor); ok && desc.IsMultiRegion() {
> ```

Shouldn't this be `!desc.IsMultiRegion()`?

Also, nit: Can we use the same pattern for initializing `querySingleRegion`. For example, in this case, the default value is "" and the `crdb_region` clause is set later whereas its the reverse above.
___
*[`pkg/sql/catalog/lease/lease.go` line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrISFlFDtnC8vS1OqZH:-NrISFlFDtnC8vS1OqZI:bf3d364) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L103)):*
> ```Go
> 			ctx, "count-leases", nil, /* txn */
> 			sessiondata.RootUserSessionDataOverride,
> 			stmt, now.GoTime(), enum.One,
> ```

we only need to pass enum.One argument when `querySingleRegion` is not an empty string, right?
___
*[`pkg/sql/catalog/lease/lease.go` line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrISydAAmrBOJse9U2N:-NrISydAAmrBOJse9U2O:bij1f8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L170)):*
> ```Go
> 			isMultiRegion = desc.IsMultiRegion()
> 		}
> 		sysDBDesc.Release(ctx)
> ```

Can we refactor this into its own library method and reuse it in `WaitForOneVersion` and `WaitForNoVersion`?



_Code quote:_
```Go
		// Detect if the system database is multi-region.
		now := m.storage.clock.Now()
		sysDBDesc, err := m.Acquire(ctx, now, keys.SystemDatabaseID)
		if err != nil {
			return nil, err
		}
		isMultiRegion := false
		if desc, ok := sysDBDesc.Underlying().(catalog.DatabaseDescriptor); ok {
			isMultiRegion = desc.IsMultiRegion()
		}
		sysDBDesc.Release(ctx)
```


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119464)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rimadeodhar)

___
*[`pkg/sql/catalog/descs/txn.go` line 71 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrIRapG8l6KDv5GzBoc:-NrIXBq1DPghd9zBeDcl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/descs/txn.go#L71)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nit: incomplete comment.
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/count.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrIRktP8Ro4K1PxzoFQ:-NrIXDCp8vL6aiyeX_Xe:brl4yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/count.go#L46)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Shouldn't this be `isMultiRegion` instead of `!isMultiRegion`? We need to add the `crdb_region` clause only for single region clusters right?
</blockquote></details>

Done. 

:facepalm:  yeah that is the wrong direction 
___
*[`pkg/sql/catalog/lease/lease.go` line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrIRwq717upG601cPDi:-NrIXRhbEnm9KyeC_Nsp:bxpdnhi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L88)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Shouldn't this be `!desc.IsMultiRegion()`?

Also, nit: Can we use the same pattern for initializing `querySingleRegion`. For example, in this case, the default value is "" and the `crdb_region` clause is set later whereas its the reverse above.
</blockquote></details>

Done.

The code is now shared.
___
*[`pkg/sql/catalog/lease/lease.go` line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrISFlFDtnC8vS1OqZH:-NrIXev00riT3HMKsVR8:b8kkfkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L103)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

we only need to pass enum.One argument when `querySingleRegion` is not an empty string, right?
</blockquote></details>

Right, there are no negative side effects otherwise, but to be safer I refactored it so a nil is passed when the query doesn't use it.
___
*[`pkg/sql/catalog/lease/lease.go` line 170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrISydAAmrBOJse9U2N:-NrIXckX7D7Y5mGcBBvl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L170)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Can we refactor this into its own library method and reuse it in `WaitForOneVersion` and `WaitForNoVersion`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (changes requested):
This is looking good. The test coverage for the backport seems low. I have requested to add a unit test. Are there any other tests within out test suite/E2E tests that you have run to ensure the countLeases method works as expected with these chages?

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-:-NrM-qDK9FkZXNLzQbZg:b5tc5y2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/catalog/lease/lease.go` line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrISFlFDtnC8vS1OqZH:-NrM-Z_f7DSIXGR6pwn0:b-69f7f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L103)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Right, there are no negative side effects otherwise, but to be safer I refactored it so a nil is passed when the query doesn't use it.
</blockquote></details>

Nice, thanks for the clarification.
___
*[`pkg/sql/catalog/lease/lease.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrM-kMC8tVLESsQ78RK:-NrM-kMC8tVLESsQ78RL:b24le4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a386d573c73c053e1a6b9127bea2625d89f5c11/pkg/sql/catalog/lease/lease.go#L78)):*
> ```Go
> // isSystemDatabaseMultiRegion returns if the system database is set-up for
> // multi-region.
> func (m *Manager) isSystemDatabaseMultiRegion(ctx context.Context) (bool, error) {
> ```

Nice, can we add a unit test for this in the PR to ensure this works as expected.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
I added a new lease_test inside the multiregion package focused on MR system databases that should validate all calls modified by this PR. This should provide a lot better coverage to make sure it works correctly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-:-NrMX6qJ3lXD2gWVCYVN:b-3i46p8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rimadeodhar)

___
*[`pkg/sql/catalog/lease/lease.go` line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrISFlFDtnC8vS1OqZH:-NrMX5Yl1Yptx71FWR6y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7aa3048841ee9a34e3c7b0966bf3058adcdb2f8a/pkg/sql/catalog/lease/lease.go#L103)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nice, thanks for the clarification.
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/lease/lease.go` line 78 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-NrM-kMC8tVLESsQ78RK:-NrMX1IADfj1X1LysiSo:b-dtls2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a386d573c73c053e1a6b9127bea2625d89f5c11/pkg/sql/catalog/lease/lease.go#L78)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

Nice, can we add a unit test for this in the PR to ensure this works as expected.
</blockquote></details>

I opted for an end to end test for multi-region, which should provide coverage for this code path.


<!-- Sent from Reviewable.io -->

rimadeodhar reviewed (approved):
Nice test! We should also add this to master. Thanks for this change.

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-:-NraogVhEHjnZQw7WNTF:blzrsv9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->

rimadeodhar commented:
We should also post this in the backports channel.

mgartner commented on pkg/sql/catalog/lease/count.go:
```diff
@@ -35,14 +51,16 @@ func CountLeases(
 		at.AsOfSystemTime()) +
-		strings.Join(whereClauses, " OR ")
+		strings.Join(whereClauses, " OR ") +
+		singleRegionClause
```
I think this is incorrect. See my comment here: https://github.com/cockroachdb/cockroach/pull/119466#pullrequestreview-1903819379

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119464#-:-NrgCzSiDs-lUPtF79bS:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->

fqazi commented:
@mgartner @rimadeodhar TFTR!
